doc/plugin_tutorial/tuto1/src/g_tuto1.mlg

```diff
@@ -16,141 +16,278 @@ open Stdarg
+
+(*** Interning terms ***)
+
+(*
+ * The next step is to make something of parsed expression.
+ * Interesting information in interp/constrintern.mli.
+ *
+ * When you read in constr(e), e will have type Constrexpr.constr_expr,
+ * which is defined in pretyping/constrexpr.ml. Your plugin
+ * will want a different representation.
+ *
+ * The important function is Constrintern.interp_constr_evars,
+ * which converts between a constr_expr and an
+ * (EConstr.constr, evar_map) pair. This essentially contains
+ * an internal representation of the term along with some state.
+ * For more on the state, read /dev/doc/econstr.md.
+ *
+ * NOTE ON INTERNING: Always prefer Constrintern.interp_constr_evars
+ * over Constrintern.interp_constr. The latter does not g
+ *
+ * To get your initial environment, call Global.env ().
+ * To get state from that environment, call Evd.from_env on that environment.
+ * It is important to NEVER use the empty environment or Evd.empty;
+ * if you do, you will get confusing errors.
+ *
+ * NOTE ON STATE: It is important to use the evar_map that is returned to you.
+ * Otherwise, you may get cryptic errors later in your plugin.
+ * For example, you may get universe inconsistency errors.
+ * In general, if a function returns an evar_map to you, that's the one
+ * you want to thread through the rest of your command.
+ *)
+VERNAC COMMAND EXTEND Intern CLASSIFIED AS QUERY
+| [ "Intern" constr(e) ] ->
+   {
+     let env = Global.env () in (* use this; never use empty *)
+     let sigma = Evd.from_env env in (* use this; never use empty *)
+     let debug sigma = Termops.pr_evar_map ~with_univs:true None env sigma in
+     Feedback.msg_notice (strbrk "State before intern: " ++ debug sigma);
+     let (sigma, t) = Constrintern.interp_constr_evars env sigma e in
+     Feedback.msg_notice (strbrk "State after intern: " ++ debug sigma);
+     let print t = Printer.pr_econstr_env env sigma t in
+     Feedback.msg_notice (strbrk "Interned: " ++ print t)
+   }
 END
 
-(* The next step is to make something of parsed expression.
-   Interesting information in interp/constrintern.mli *)
-
-(* There are several phases of transforming a parsed expression into
-  the final internal data-type (constr).  There exists a collection of
-  functions that combine all the phases *)
-
-VERNAC COMMAND EXTEND TakingConstr2 CLASSIFIED AS QUERY
-| [ "Cmd2" constr(e) ] ->
-  { let _ = Constrintern.interp_constr
-    (Global.env())
-    (* Make sure you don't use Evd.empty here, as this does not
-      check consistency with existing universe constraints. *)
-    (Evd.from_env (Global.env())) e in
-    Feedback.msg_notice (strbrk "Cmd2 parsed something legitimate") }
+(*** Defining terms ***)
+
+(*
+ * To define a term, we start similarly to our intern functionality,
+ * then we call another function. We define this function in
+ * the Simple_declare module.
+ *
+ * The line #[ poly = polymorphic ] says that this command accepts
+ * polymorphic attributes.
```

Comment:
Here `polymorphic` is `Attributes.polymorphic` (resolved because we `open Attributes` earlier in the file) and `poly` is what the result is bound to in the rule's code. Multiple attributes may be used separated by `;`, and we have punning so `foo` is equivalent to `foo = foo`.